include/ccf/js/registry.h

```diff
@@ -127,6 +131,28 @@ namespace ccf::js
     ccf::ApiResult get_js_runtime_options_v1(
       ccf::JSRuntimeOptions& options, kv::ReadOnlyTx& tx);
 
+    /**
+     * Record action details by storing them in KV maps using a common format,
+     * for the purposes of offline audit using the ledger.
+     */
+    ccf::ApiResult record_action_details_for_audit_v1(
+      kv::Tx& tx,
+      ccf::ActionFormat format,
+      const std::string& user_id,
+      const std::string& action_name,
+      const std::vector<uint8_t>& action_body);
+
+    /**
+     * Check an action is original, as opposed to replayed, by looking it up
+     * in the history of recent actions. To place an upper bound on the history
+     * size, an authenticated timestamp (@p created_at) is required.
+     */
+    ccf::ApiResult is_original_action_execution_v1(
```

Comment:
`nonreplayed` is not really a word though, how about `check_action_not_replayed_v1()`. I prefer to avoid negatives in API names, because they lead to double negatives which can be confusing, but with a non-boolean return, that's less of a problem.

I'd rather keep the API returns uniform as much as possible, this is not a compelling case for breaking that consistency.